Add three/addons target for NPM.

CodyJasonBennett opened the pull request:
Related issue: #24595

**Description**

Tries to pick up after #23368 which was closed due to lack of tree-shaking which we've found unnecessary with #26912.

CodyJasonBennett committed:
Add three/addons target

CodyJasonBennett committed:
Update Addons.js

LeviPesin commented on examples/jsm/Addons.js:
```diff
@@ -0,0 +1,492 @@
+export * from './modifiers/EdgeSplitModifier.js';
+export * from './modifiers/SimplifyModifier.js';
+export * from './modifiers/TessellateModifier.js';
+export * from './nodes/Nodes.js';
```
No more exports are needed for the Nodes system, this line is sufficient.

CodyJasonBennett replied:
I'm wondering if we should exclude WebGPU (and capabilities/WebGPU) also with #26626 for the moment.

marcofugaro replied:
Yeah I think the WebGPU/Nodes should not be included into the `three/addons` for now, their tree-shaking is more complicated and WebGPU at some point will be moved into core.

CodyJasonBennett committed:
Only export Nodes entry

TODO: this maybe should be removed https://github.com/mrdoob/three.js/pull/26910#discussion_r1349487036

CodyJasonBennett committed:
Addons: don't export WebGPU, nodes, offscreen

CodyJasonBennett commented on examples/jsm/Addons.js:
```diff
@@ -0,0 +1,334 @@
+export * from './objects/Sky.js';
+export * from './objects/Water.js';
+export * from './objects/Water2.js';
+// export * from './offscreen/jank.js';
+// export * from './offscreen/offscreen.js';
+// export * from './offscreen/scene.js';
```
I'm not sure why these exist or how they should be exported. Can't find any references in the repo.

marcofugaro replied:
You can ignore them, they are used in this example: https://threejs.org/examples/#webgl_worker_offscreencanvas

These are not reusable addons and it's logic relative to that example, doesn't make sense for them to be there. Ideally they should be inlined in the example, as an inline web worker.

CodyJasonBennett committed:
Addons: resolve conflicts, add namespace aliases

CodyJasonBennett committed:
Addons: remove offscreen, WIP nodes/wgpu

marcofugaro commented on examples/jsm/Addons.js:
```diff
@@ -0,0 +1,292 @@
+export * from './interactive/SelectionBox.js';
+export * from './interactive/SelectionHelper.js';
+
+export * from './lights/IESSpotLight.js';
```
this files contains a default export, maybe change it to named export like any other addons

marcofugaro commented on examples/jsm/Addons.js:
```diff
@@ -0,0 +1,292 @@
+export * from './loaders/TDSLoader.js';
+export * from './loaders/TGALoader.js';
+export * from './loaders/TIFFLoader.js';
+export * from './loaders/TTFLoader.js';
```
the `opentype.module.js` breaks esbuild with default settings when importing anything from `three/addons`. 

<img width="1066" alt="Screenshot 2023-10-16 at 11 51 25" src="https://github.com/mrdoob/three.js/assets/7217420/3c33adc8-dee8-429a-8882-40ebb7c92f36">

How should we handle this case?

CodyJasonBennett replied:
We can remove the Node code (e.g. https://github.com/pmndrs/three-stdlib/commit/9753e6b07890ae393564b3ef205104c7461db6f7). It's only used for `Font.prototype.download` which isn't called by three.js code.

marcofugaro commented on examples/jsm/utils/BufferGeometryUtils.js:
```diff
@@ -1359,6 +1359,23 @@ function mergeBufferAttributes( attributes ) {
 
 }
 
+const BufferGeometryUtils = {
+
+	computeMikkTSpaceTangents,
+	mergeGeometries,
+	mergeBufferGeometries,
+	mergeAttributes,
+	mergeBufferAttributes,
+	interleaveAttributes,
+	estimateBytesUsed,
+	mergeVertices,
+	toTrianglesDrawMode,
+	computeMorphedAttributes,
+	mergeGroups,
+	toCreasedNormals,
+
+};
```
This approach doesn't allow for tree-shaking inside the `BufferGeometryUtils.js` file, meaning that if you only use one function, you're gonna have to include the whole file.

You sure we can't use `export * as` yet?
This is not the main bundle, and CRA 4 users would have the old way of importing addons as an alternative as well.

CodyJasonBennett replied:
That's enough reason for me. FWIW namespaced exports are ES2020 or in browsers since early 2021 https://caniuse.com/mdn-javascript_statements_export_namespace.

CodyJasonBennett committed:
opentype: remove Node-specific methods

CodyJasonBennett committed:
Addons: prefer namespace export, handle IESSpotLight default export

marcofugaro commented:
Alright, this PR looks good to me 👍 ready to be merged @mrdoob @Mugen87 

After further testing, we found out that this PR is already tree-shakeable by iteself and doesn't need the changes from https://github.com/mrdoob/three.js/pull/26912.

CodyJasonBennett committed:
Merge branch 'dev' into npm/addons-target

CodyJasonBennett commented:
Thanks to @marcofugaro for the sanity check and for championing this as much as me.
